pkg/agent/flowexporter/connections/l7_listener.go

```diff
@@ -0,0 +1,221 @@
+	}
+	defer listener.Close()
+
+	klog.InfoS("L7 Listener Server started. Listening for connections...")
+	for {
+		select {
+		case <-stopCh:
+			return
+		default:
+			conn, err := listener.Accept()
+			if err != nil {
+				klog.ErrorS(err, "Error accepting Suricata connection")
+				continue
+			}
+			go l.handleClientConnection(conn)
+		}
+	}
+}
+
+func (l *L7Listener) handleClientConnection(conn net.Conn) {
+	defer conn.Close()
+	reader := bufio.NewReader(conn)
+	for {
+		buffer, err := reader.ReadBytes('\n')
+		if err != nil && err != io.EOF {
+			klog.ErrorS(err, "Error reading data", "buffer", buffer)
+			return
+		}
+		err = l.processLog(buffer)
+		if err != nil {
+			klog.ErrorS(err, "Error while processing L7 data")
+			return
+		}
+	}
+}
+
+func (l *L7Listener) processLog(data []byte) error {
+	var event JsonToEvent
+	err := json.Unmarshal(data, &event)
+	if err != nil {
+		return fmt.Errorf("error parsing JSON data %v", data)
+	}
+	if event.EventType != "http" {
+		return nil
+	}
+	if err = l.addOrUpdateL7EventMap(&event); err != nil {
+		return fmt.Errorf("error while adding or updating L7 event map %v", err)
+	}
+	return nil
+}
+
+func (l *L7Listener) addOrUpdateL7EventMap(event *JsonToEvent) error {
+	l.l7mut.Lock()
+	defer l.l7mut.Unlock()
+	protocol, err := flowexporter.LookupProtocolMap(event.Proto)
+	if err != nil {
+		return fmt.Errorf("invalid protocol type, err: %v", err)
+	}
+	conn := flowexporter.Connection{
+		FlowKey: flowexporter.Tuple{
+			SourceAddress:      event.SrcIP,
+			DestinationAddress: event.DestIP,
+			Protocol:           protocol,
+			SourcePort:         uint16(event.SrcPort),
+			DestinationPort:    uint16(event.DestPort),
+		},
+	}
+	connKey := flowexporter.NewConnectionKey(&conn)
+	srcIP := conn.FlowKey.SourceAddress.String()
+	dstIP := conn.FlowKey.DestinationAddress.String()
+	startTime, _ := time.Parse(time.RFC3339Nano, event.Timestamp)
+	srcPod, srcFound := l.podStore.GetPodByIPAndTime(srcIP, startTime)
+	dstPod, dstFound := l.podStore.GetPodByIPAndTime(dstIP, startTime)
+	if !srcFound && !dstFound {
+		return fmt.Errorf("Cannot map any of the IP %s or %s to a local Pod", srcIP, dstIP)
+	}
+	var sourcePodNN, destinationPodNN string
+	if srcFound {
+		sourcePodNN = k8sutil.NamespacedName(srcPod.Namespace, srcPod.Name)
+	}
+	if dstFound {
+		destinationPodNN = k8sutil.NamespacedName(dstPod.Namespace, dstPod.Name)
+	}
+
+	switch event.EventType {
+	case "http":
+		if l.podL7VisibilityAttrGetter.IsL7VisibilityRequested(sourcePodNN, false) || l.podL7VisibilityAttrGetter.IsL7VisibilityRequested(destinationPodNN, true) {
+			if _, ok := l.l7Events[connKey]; !ok {
+				l.l7Events[connKey] = make(map[string]map[int32]Http)
+				l.l7Events[connKey]["http"] = make(map[int32]Http)
+			}
+			l.l7Events[connKey]["http"][event.TxID] = event.HTTP
+		}
+	}
+	return nil
+}
+
+func (l *L7Listener) CopyL7EventMap() map[flowexporter.ConnectionKey]map[string]map[int32]Http {
```

Comment:
we have a triple-nested map as the type here, and throughout the code base

would it make sense for readability to define a new type (with all the appropriate methods):
```
type L7EventMap struct {
        // m maps the protocol (e.g., "http") to a map indexed by the transaction ID
        m map[string]map[int32]Http
}
```

there is one more thing to point out: the "middle" map is indexed by protocol (with "http" the only supported value at the moment), yet the final value type is "hardcoded" to `Http`. Let's assume we add support for a new protocol (e.g., "ssh"), what would the map look like?